### PR TITLE
Refactor moveit_servo launch testing

### PR DESCRIPTION
**Pull request by tylerjw**
### Description

This change is taking into account what we learned here about launching gtests with parameters:  https://github.com/ros2/launch/issues/466

This removes the workaround that @AdamPettinger found and makes the tests depend directly on the parameter loading.  This is useful because there are no longer two copies of the config, therefore the test that compared the struct version with the loaded yaml version is no longer necessary.

I also broke the parameter loading into a separate library so it could be included by tests that needed the parameters but otherwise did not have a reason to link in the servo library itself.

Next, I made the parameters a shared_ptr to a constant struct.  The advantage of this is the params are no longer mutable during runtime.  This reduces the number of mutable variables and therefore the number of things the reader has to keep track of when trying to grok the source.  This especially useful for reasoning about immutability when debugging threading issues.  One side-effect of this is the unit test for servo_cals that does modify the parameter struct during runtime has to now use const_cast.  This is a reasonable thing to do in a test imo.

Lastly, because the struct is gone the unit test for servo_calcs has to be launched with a launch file because it needs the parameters to be loaded for it.

I hope that this can be a good example of how to write gtests (and their launch files) for both integration and unit tests, and how to load yaml parameters into those tests.  My primary motivation in doing this test was to learn about how the launch system works for gtests in ros2.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [x] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [x] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

### CODE REVIEW

File: moveit_ros/moveit_servo/test/servo_launch_test_common.hpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/test/servo_launch_test_common.hpp:91]: @henningkayser here is a way to load parameters into const storage.  This uses the shared_ptr to a const as a class variable: `servo_parameters_` and a local non-const one.  The parameters are loaded into the temporary non-const parameters and then the shared pointer is given the ownership and the local non-const goes out of scope.

File: moveit_ros/moveit_servo/test/unit_test_servo_calcs.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/test/unit_test_servo_calcs.cpp:539]: Here I am also doing the parameter initialization, just this time it is in a main, and when it fails I can log and terminate.  In the case above where it is in a constructor, the workaround is I have to store a variable for the return on readParmeters and later fail if that variable is false.  Generally, I think we should keep reading parameters out of constructors except in the case of test fixtures.

**Comment by tylerjw** [comment]
The reason this is failing is an initialization timing issue.  The integration test launches servo the same way the demo does with a component container and then the gtest sits around waiting for the start service to be registered.  If you read the logs for some reason the servo component never gets loaded.  I think we should find some way in the launch file to launch the gtest node only after the servo component gets loaded.  I posted a question to ros answers here about this: https://answers.ros.org/question/370324/launch-node-when-component-node-is-loaded/

The issue with this PR exists outside of this PR but I'm ok waiting on this until we figure out the answer to how to launch the test when a component node loads.

File: moveit_ros/moveit_servo/src/collision_check.cpp
Comment (henningkayser) [inline moveit_ros/moveit_servo/src/collision_check.cpp:56]: Not sure if this renaming is really worth it, it touches a lot of lines and will definitely make syncing difficult in the future.

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (henningkayser) [inline moveit_ros/moveit_servo/src/servo_parameters.cpp:90]: no default? Then I would remove this comment

File: moveit_ros/moveit_servo/test/servo_launch_test_common.hpp
Comment (henningkayser) [reply moveit_ros/moveit_servo/test/servo_launch_test_common.hpp:91]: Nice solution! I had something similar in mind (more inside the struct for const/non-const parameters) but this is probably the way to go for the beginning.

File: moveit_ros/moveit_servo/test/test_servo_interface.cpp
Comment (henningkayser) [inline moveit_ros/moveit_servo/test/test_servo_interface.cpp:168]: nit: using seconds would give you a `duration` in exchange for a `1000 *` and a `milli`. Or maybe even nicer to read would be `rclcpp::sleep_for(1.5s *  servo_parameters_->incoming_command_timeout);`...

File: moveit_ros/moveit_servo/test/unit_test_servo_calcs.cpp
Comment (henningkayser) [reply moveit_ros/moveit_servo/test/unit_test_servo_calcs.cpp:539]: very useful example

**Review by henningkayser — APPROVED** [review]
I added some suggestions for improvements, but in general I approve these changes

**Review by AndyZe — CHANGES_REQUESTED** [review]
Given that Servo has been broken for some time (in MoveIt2), I would like to start testing all changes. Blocking until I have time for that.

File: moveit_ros/moveit_servo/CMakeLists.txt
Comment (AndyZe) [inline moveit_ros/moveit_servo/CMakeLists.txt:4]: Please add a comment for what each of these is. Remember that to most people, it's hard to tell the difference between `moveit_servo` and `servo_server`. I think that should be:

```
# Core Servo library, for calculations and collision checking. Provides a C++ interface
set(SERVO_LIB_NAME moveit_servo)
# Library for parameter loading
set(SERVO_PARAM_LIB_NAME ${SERVO_LIB_NAME}_parameters)
# A library to run as a ROS node component, and receive commands via topics
set(COMPOSABLE_NODE_NAME servo_server)
```

File: moveit_ros/moveit_servo/CMakeLists.txt
Comment (AndyZe) [inline moveit_ros/moveit_servo/CMakeLists.txt:56]: Let's specify that it's a C++ interface, not a ROS2 topic interface:

`# This core library provides a C++ interface for sending realtime twist or joint commands to a robot`

File: moveit_ros/moveit_servo/CMakeLists.txt
Comment (AndyZe) [inline moveit_ros/moveit_servo/CMakeLists.txt:68]: `# Add and export library to run as a ROS node component, and receive commands via topics`

File: moveit_ros/moveit_servo/CMakeLists.txt
Comment (AndyZe) [inline moveit_ros/moveit_servo/CMakeLists.txt:190]: Why not export `COMPOSABLE_NODE_NAME` too? This is the library that provides a ROS node component

File: moveit_ros/moveit_servo/doc/servo_tutorial.md
Comment (AndyZe) [inline moveit_ros/moveit_servo/doc/servo_tutorial.md:38]: nit: `standalone node component that receives commands as ROS topics`

File: moveit_ros/moveit_servo/doc/servo_tutorial.md
Comment (AndyZe) [inline moveit_ros/moveit_servo/doc/servo_tutorial.md:72]: This is actually outdated now. `position_controllers/JointGroupPositionControllers` and `velocity_controllers/JointGroupVelocityControllers` are available. Suggest simply deleting this line.

**Comment by AndyZe** [comment]
This would close #306

File: moveit_ros/moveit_servo/include/moveit_servo/servo.h
Comment (AndyZe) [inline moveit_ros/moveit_servo/include/moveit_servo/servo.h:92]: Why delete this?

File: moveit_ros/moveit_servo/src/collision_check.cpp
Comment (AndyZe) [reply moveit_ros/moveit_servo/src/collision_check.cpp:56]: yeah, i tend to agree

File: moveit_ros/moveit_servo/test/servo_launch_test_common.hpp
Comment (AndyZe) [reply moveit_ros/moveit_servo/test/servo_launch_test_common.hpp:91]: Can you add a further comment at L86.5 about why that matters? For those of us who are new to ROS2   ;)

Maybe something like:

`The constant can be shared by many individual node components`

File: moveit_ros/moveit_servo/test/servo_launch_test_common.hpp
Comment (AndyZe) [inline moveit_ros/moveit_servo/test/servo_launch_test_common.hpp:476]: The ROS2 style guide refers to the Google style guide here. It says the variable names should be camel_case.

https://google.github.io/styleguide/cppguide.html#Variable_Names

**Comment by AndyZe** [comment]
Overall I really like the simplifications here. I tested that `servo_cpp_interface_demo.launch.py` still works. Hopefully both demos work after rebasing.

File: moveit_ros/moveit_servo/CMakeLists.txt
Comment (tylerjw) [reply moveit_ros/moveit_servo/CMakeLists.txt:4]: What do you think of calling the node just `servo` or `servo_node`?  The word `server` doesn't mean much in ros and seems confusing to me.

File: moveit_ros/moveit_servo/src/collision_check.cpp
Comment (tylerjw) [reply moveit_ros/moveit_servo/src/collision_check.cpp:56]: Ok, I'll remove the rename.  I thought it would make it clearer because there is a pose tracker parameters struct also.  I agree that this is unrelated to this PR and shouldn't be here.

File: moveit_ros/moveit_servo/src/servo_parameters.cpp
Comment (tylerjw) [reply moveit_ros/moveit_servo/src/servo_parameters.cpp:90]: I think it is nice to have the default args as comments in the code when they are now specified in a header file.  If you have a function that's defined in a header with default args you can't also specify them in the source file so this is the tactic I've used to show when looking at the code that the parameter has a default value.

File: moveit_ros/moveit_servo/test/test_servo_interface.cpp
Comment (tylerjw) [reply moveit_ros/moveit_servo/test/test_servo_interface.cpp:168]: I agree, I'll clean this up in a follow-on PR.  I am going to remove the rename.

File: moveit_ros/moveit_servo/include/moveit_servo/servo.h
Comment (tylerjw) [reply moveit_ros/moveit_servo/include/moveit_servo/servo.h:92]: I was going to replace this function call with a service (so it could be done through the component interface also).  However, I don't think that change belongs in this PR.  I'll put this back.

File: moveit_ros/moveit_servo/test/servo_launch_test_common.hpp
Comment (tylerjw) [reply moveit_ros/moveit_servo/test/servo_launch_test_common.hpp:476]: You are right.  I named these this way to make it clear that they are constants everywhere they are used.  I agree though that this is not following the style guide and I will fix it.

File: moveit_ros/moveit_servo/CMakeLists.txt
Comment (tylerjw) [reply moveit_ros/moveit_servo/CMakeLists.txt:4]: Or maybe even servo_component?

Comment (AndyZe) [reply moveit_ros/moveit_servo/CMakeLists.txt:4]: i like `servo_component`  :+1:

File: moveit_ros/moveit_servo/CMakeLists.txt
Comment (tylerjw) [reply moveit_ros/moveit_servo/CMakeLists.txt:190]: The way I'm using it here is even wrong (these shouldn't be library names).  Here is the issue I created about my confusion around how to use this and the official documentation.  https://github.com/ros2/ros2_documentation/issues/1094

The short of it is this line should not be here and the correct (as far as I understand now) should be this instead:

```cmake
# Install Targets
install(
  TARGETS
    ${SERVO_LIB_NAME}
    ${SERVO_PARAM_LIB_NAME}
    ${COMPONENT_NODE_NAME}
    ${SERVO_CONTROLLER_INPUT}
    ${CPP_DEMO_NAME}
    ${FAKE_SERVO_CMDS_NAME}
  EXPORT
    export_${PROJECT_NAME}
  ARCHIVE DESTINATION lib
  LIBRARY DESTINATION lib
  RUNTIME DESTINATION bin
  INCLUDES DESTINATION include
)

# This calls the install(EXPORT ...) cmake functions to export the target libraries
# so other pojects can depend on libraries from moveit_servo.
ament_export_targets(export_${PROJECT_NAME} HAS_LIBRARY_TARGET)
ament_export_dependencies(${THIS_PACKAGE_INCLUDE_DEPENDS})
```

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:1134]: @AndyZe This is why I removed this function.  This is no longer a constant parameter.  This parameter needs to be handled separately from the rest of the parameters if it can be changed during runtime from the rest of the parameters.

What if this changes between the use of it here: https://github.com/ros-planning/moveit2/blob/bbe859ee6c2bd35c9ba01f666ff9e9a8a4915e1f/moveit_ros/moveit_servo/src/servo_calcs.cpp#L309-L314 and here: https://github.com/ros-planning/moveit2/blob/bbe859ee6c2bd35c9ba01f666ff9e9a8a4915e1f/moveit_ros/moveit_servo/src/servo_calcs.cpp#L463

Should the value of this be changed only between runs of the servo calcs loop or can it be changed while it is being used?

**Comment by tylerjw** [comment]
@AndyZe @henningkayser I responded to your reviews and am now rebasing against main.  I removed the renames and will make separate PRs for them.  I made my response to comments a separate commit so it would be easy to look at it separately however I expect to squash this before landing it.

File: moveit_ros/moveit_servo/include/moveit_servo/servo.h
Comment (AndyZe) [reply moveit_ros/moveit_servo/include/moveit_servo/servo.h:92]: I'm glad you commented it. This is a useful function that I would like to keep.

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (AndyZe) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:1212]: ideally it could be changed at any time, not just while Servo is paused

File: moveit_ros/moveit_servo/test/launch/unit_test_servo_calcs.test.py
Comment (AndyZe) [inline moveit_ros/moveit_servo/test/launch/unit_test_servo_calcs.test.py:26]: nit: what is this?

**Review by AndyZe — APPROVED** [review]

File: moveit_ros/moveit_servo/test/launch/unit_test_servo_calcs.test.py
Comment (tylerjw) [reply moveit_ros/moveit_servo/test/launch/unit_test_servo_calcs.test.py:26]: This was leftover from me debugging, I can remove it if you'd like.  It is in the test launch and is useful to run gdb in a separate window.  I can remove it or add a comment about it, I don't care either way.

**Comment by tylerjw** [comment]
@AndyZe sorry to keep adding to this PR.  I didn't think it would be good to merge this removing the functionality of `changeRobotLinkCommandFrame` so I went and learned about how the dynamic parameter system works in ROS2.  This should be a good example for how to use it.

@henningkayser take a look at the last commit I just added to this.  It is my idea for how to add dynamic parameter handling to a parameter struct.

Another interesting thing is I realized the parameter builder struct could return a shared_ptr<const and that would guarantee it is used the way I expect.

The only thing I don't like about this dynamic parameter interface is the object that cares about the dynamic parameter has to handle the state of the parameter instead of the parameter struct handling it.  If I could have some more time to think about this maybe I could find another way around that.

**Comment by tylerjw** [comment]
I refactored this and it is now passing reliably locally and I can no longer reproduce the segfault (let's hope it was an upstream bug that is now fixed).

**Pull request merged**
